packages/dds/tree/src/core/edit-manager/editManager.ts

```diff
@@ -113,119 +166,141 @@ export class EditManager<
+			const { change } =
+				localPath.shift() ??
 				fail(
 					"Received a sequenced change from the local session despite having no local changes",
 				);
-
-			// TODO: The local change may contain references to local revision tags.
-			// When other clients rebase this change, they will instead use the corresponding sequence numbers
-			// instead of this client's local revision numbers. Although this need not inherently cause a divergence
-			// in the state of the document, this is potentially dangerous.
-			// This could be solved by one of the following approaches:
-			// A) Maintain a `Branch` of sequenced changes from this client in addition to `localChanges`,
-			//    and only use the rebased changes from that branch when adding to the trunk.
-			// B) Use a UUID as the revision tag for each changeset instead of a sequence number,
-			//    so that we do not need local revision tags.
-			// C) Strip any references to local changes before adding the change to the trunk.
-			//    This would also have to be done on each branch with references to changes local to that branch.
-			this.trunk.push({
-				...newCommit,
-				changeset: change.change,
-			});
+			this.pushToTrunk(sequenceNumber, { ...newCommit, change });
+			// TODO: Can this be optimized by simply mutating the localPath parent pointers? Is it safe to do that?
+			this.localBranch = localPath.reduce(mintCommit, this.trunk);
 			return emptyDelta;
 		}
 
-		const branch = this.getOrCreateBranch(newCommit.sessionId, newCommit.refNumber);
+		// Get the revision that the remote change is based on
+		const baseRevisionInTrunk =
+			this.sequenceMap.getPairOrNextLower(referenceSequenceNumber)?.[1] ?? this.trunkBase;
 
-		// Like `newCommit.refNumber` but takes into account cases where the commit comes from a peer whose
-		// prior changes made it to the trunk without requiring any rebasing.
-		const effectiveNewCommitRef: SeqNumber = brand(
-			Math.max(branch.refSeq, newCommit.refNumber),
+		// Rebase that branch over the part of the trunk up to the base revision
+		// This will be a no-op if the sending client has not advanced since the last time we received an edit from it
+		const [rebasedBranch] = this.rebaser.rebaseBranch(
+			getOrCreate(this.peerLocalBranches, newCommit.sessionId, () => baseRevisionInTrunk),
+			this.trunk,
+			baseRevisionInTrunk,
 		);
 
-		this.updateBranch(branch, effectiveNewCommitRef);
-		const newChangeFullyRebased = this.rebaseChangeFromBranchToTrunk(newCommit, branch);
-		if (
-			this.trunk.length > 0 &&
-			effectiveNewCommitRef !== this.trunk[this.trunk.length - 1].seqNumber
-		) {
-			// The change diverges from the trunk
-			branch.localChanges.push(newCommit);
+		if (rebasedBranch === this.trunk) {
+			// If the branch is fully caught up and empty after being rebased, then push to the trunk directly
+			this.pushToTrunk(sequenceNumber, newCommit);
+			this.peerLocalBranches.set(newCommit.sessionId, this.trunk);
```

Comment:
Can we just delete the peer branch instead?

Reply:
Yea!!!!!!